pkg/splunk/enterprise/util.go

```diff
@@ -292,3 +342,53 @@ func DeleteOwnerReferencesForS3SecretObjects(client splcommon.ControllerClient,
 
 	return err
 }
+
+// GetAppListFromS3Bucket gets the list of apps from remote storage.
+func GetAppListFromS3Bucket(client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterprisev1.AppFrameworkSpec) (map[string]splclient.S3Response, error) {
+
+	scopedLog := log.WithName("GetAppListFromS3Bucket").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	sourceToAppListMap := make(map[string]splclient.S3Response)
+
+	volMap := make(map[string]enterprisev1.VolumeSpec)
+
+	for _, vol := range appFrameworkRef.VolList {
+		volMap[vol.Name] = vol
+	}
+
+	scopedLog.Info("Getting the list of apps from remote storage...")
+
+	var s3Response splclient.S3Response
+	var vol enterprisev1.VolumeSpec
+	var err error
+	var s3Client splclient.S3Client
+
+	for _, appSource := range appFrameworkRef.AppSources {
+		// get the volume spec from the volume name
+		if appSource.AppSourceDefaultSpec.VolName != "" {
+			vol = volMap[appSource.AppSourceDefaultSpec.VolName]
+		} else {
+			vol = volMap[appFrameworkRef.Defaults.VolName]
+		}
+
+		//get the corresponding S3 client
+		s3Client, err = GetRemoteStorageClient(client, cr, appFrameworkRef, &vol, appSource.Location)
+		if err != nil {
+			// move on to the next appSource if we are not able to get the rquired client
+			scopedLog.Error(err, "Unable to get remote storage client", "appSource", appSource.Name)
+			continue
+		}
+
+		// Now, get the apps list from remote storage
+		s3Response, err = s3Client.GetAppsList()
+		if err != nil {
+			// move on to the next appSource if we are not able to get apps list
+			scopedLog.Error(err, "Unable to get apps list", "appSource", appSource.Name)
+			continue
+		}
```

Comment:
Wondering if we need to have a retry OR a carry-forward retry based on some flag?, otherwise, silently returning from this function can't trigger the retry. Can be an issue with a large probing interval.s

Reply:
So I will be checking the size of `sourceToAppListMap` returned from here and the size of appSources, if they are not same, we will reconcile. So, I think the reconcile will take care of the retry mechanism itself, right?